File: daphne-server/srcs/server_controller/timesync.cpp
#include "server_controller/timesync.hpp"
#include <arpa/inet.h>
#include <algorithm>
#include <limits>
#include <mutex>
#include <regex>
#include <stdexcept>
#include <system_error>

namespace daphne_sc {
namespace {
void need(bool good) { if (!good) throw std::runtime_error("Invalid timesync observation"); }
bool owner(const std::string& value) { return value.size() <= 80 && std::regex_match(value, std::regex(":[0-9]+(?:\\.[0-9]+)+")); }
bool name(const std::string& value) {
  if (value.empty()) return true;
  if (value.size() > 253) return false;
  unsigned char address[16];
  if (inet_pton(AF_INET, value.c_str(), address) == 1 || inet_pton(AF_INET6, value.c_str(), address) == 1) return true;
  std::string text = value;
  if (text.back() == '.') text.pop_back();
  if (text.empty()) return false;
  size_t start = 0;
  for (;;) {
    const auto end = text.find('.', start);
    const auto label = text.substr(start, end == std::string::npos ? end : end - start);
    if (label.empty() || label.size() > 63 || !std::regex_match(label, std::regex("[A-Za-z0-9](?:[A-Za-z0-9-]*[A-Za-z0-9])?"))) return false;
    if (end == std::string::npos) return true;
    start = end + 1;
  }
}
bool address(const std::string& value) {
  unsigned char bytes[16];
  return value.empty() || (value.size() <= 45 &&
      (inet_pton(AF_INET, value.c_str(), bytes) == 1 || inet_pton(AF_INET6, value.c_str(), bytes) == 1));
}
bool same_context(const TimesyncRaw& a, const TimesyncRaw& b) {
  return a.bus_id == b.bus_id && a.owner_before == b.owner_before &&
      a.selected_name == b.selected_name && a.selected_address == b.selected_address;
}
bool same_sample(const NtpSampleRaw& a, const NtpSampleRaw& b) {
  return a.leap == b.leap && a.version == b.version && a.mode == b.mode && a.stratum == b.stratum &&
      a.precision == b.precision && a.root_delay_us == b.root_delay_us && a.root_dispersion_us == b.root_dispersion_us &&
      a.timestamps_us == b.timestamps_us && a.ignored_spike == b.ignored_spike && a.jitter_us == b.jitter_us;
}
daphne::NtpSampleObservation decode(const NtpSampleRaw& raw) {
  daphne::NtpSampleObservation result;
  result.set_age_bound_quality(daphne::MEASUREMENT_UNAVAILABLE);
  if (!raw.count) {
    result.set_quality(daphne::MEASUREMENT_UNAVAILABLE);
    result.set_detail("No processed NTP sample reported; zero-filled daemon state is not a measurement");
    return result;
  }
  try {
    need(raw.leap <= 2 && (raw.version == 3 || raw.version == 4) && raw.mode == 4 && raw.stratum > 0 && raw.stratum < 16);
    need(raw.precision >= -128 && raw.precision <= 127);
    for (auto stamp : raw.timestamps_us) need(stamp > 0 && stamp <= static_cast<uint64_t>(INT64_MAX) / 1000);
    const auto t1 = static_cast<__int128>(raw.timestamps_us[0]), t2 = static_cast<__int128>(raw.timestamps_us[1]);
    const auto t3 = static_cast<__int128>(raw.timestamps_us[2]), t4 = static_cast<__int128>(raw.timestamps_us[3]);
    const auto delay = (t4 - t1) - (t3 - t2);
    const auto offset = ((t2 - t1) + (t3 - t4)) * 500; // Preserve half-microsecond units exactly.
    need(t4 >= t1 && t3 >= t2 && delay >= 0 && delay <= static_cast<__int128>(UINT64_MAX) / 1000);
    need(offset >= INT64_MIN && offset <= INT64_MAX);
    result.set_leap(raw.leap); result.set_version(raw.version); result.set_mode(raw.mode); result.set_stratum(raw.stratum);
    result.set_precision_exponent(raw.precision);
    result.set_root_delay_us(raw.root_delay_us); result.set_root_dispersion_us(raw.root_dispersion_us);
    result.set_origin_unix_us(raw.timestamps_us[0]); result.set_receive_unix_us(raw.timestamps_us[1]);
    result.set_transmit_unix_us(raw.timestamps_us[2]); result.set_destination_unix_us(raw.timestamps_us[3]);
    result.set_ignored_spike(raw.ignored_spike); result.set_jitter_us(raw.jitter_us);
    result.set_offset_ns(static_cast<int64_t>(offset)); result.set_round_trip_delay_ns(static_cast<uint64_t>(delay) * 1000);
    result.set_quality(daphne::MEASUREMENT_GOOD);
    result.set_detail("Historical processed response; ignored-spike flag retained. Not a present offset, successful adjustment, authenticated peer or fresh UTC guarantee");
  } catch (const std::exception&) {
    result.Clear();
    result.set_quality(daphne::MEASUREMENT_ERROR);
    result.set_age_bound_quality(daphne::MEASUREMENT_UNAVAILABLE);
    result.set_detail("Invalid historical NTP sample; fields withheld");
  }
  return result;
}
}  // namespace

void TimesyncHistory::reset() { prior_.reset(); prior_start_ = prior_end_ = 0; not_before_.reset(); }
std::optional<uint64_t> TimesyncHistory::observe(const TimesyncRaw& raw, uint64_t start, uint64_t end) {
  need(start && end >= start);
  if (!prior_ || !same_context(*prior_, raw) || start < prior_end_ || raw.sample.count < prior_->sample.count) {
    not_before_.reset();
  } else if (raw.sample.count > prior_->sample.count) {
    not_before_ = prior_start_; // New response was processed after the preceding snapshot read.
  } else if (!same_sample(raw.sample, prior_->sample)) {
    // Equal count but changed payload does not establish a stable sample epoch.
    not_before_.reset();
  }
  prior_ = raw; prior_start_ = start; prior_end_ = end;
  return raw.sample.count ? not_before_ : std::nullopt;
}

daphne::TimesyncObservation read_timesync(TimesyncIo& io, TimesyncHistory& history, bool private_details) {
  daphne::TimesyncObservation result;
  result.set_source("local system bus:org.freedesktop.timesync1.Manager:GetAll");
  try {
    const auto start = io.monotonic_ns();
    result.set_acquisition_started_monotonic_ns(start);
    need(start && start <= UINT64_MAX - kTimesyncMaximumCollectionNs);
    const auto raw = io.query(start + kTimesyncMaximumCollectionNs);
    const auto end = io.monotonic_ns();
    need(end >= start && end - start <= kTimesyncMaximumCollectionNs);
    need(std::regex_match(raw.bus_id, std::regex("[0-9a-f]{32}")) && owner(raw.owner_before));
    need(raw.owner_before == raw.owner_after && raw.owner_before == raw.reply_sender);
    need(name(raw.selected_name) && address(raw.selected_address));
    need(raw.poll_min_us > 0 && raw.poll_max_us >= raw.poll_min_us && raw.root_max_us > 0);
    need(raw.poll_us == 0 || (raw.poll_us >= raw.poll_min_us && raw.poll_us <= raw.poll_max_us));
    result.set_bus_id(raw.bus_id); result.set_unique_owner(raw.owner_before);
    result.set_owner_bracket_verified(true); result.set_details_included(private_details);
    result.set_selected_name_present(!raw.selected_name.empty());
    result.set_selected_address_present(!raw.selected_address.empty());
    if (private_details && !raw.selected_name.empty()) result.set_selected_server_name(raw.selected_name);
    if (private_details && !raw.selected_address.empty()) result.set_selected_server_address(raw.selected_address);
    result.set_poll_interval_us(raw.poll_us); result.set_poll_minimum_us(raw.poll_min_us); result.set_poll_maximum_us(raw.poll_max_us);
    result.set_root_distance_maximum_us(raw.root_max_us); result.set_frequency_scaled_ppm(raw.frequency_scaled_ppm);
    result.set_processed_packet_count(raw.sample.count);
    *result.mutable_last_sample() = decode(raw.sample);
    if (result.last_sample().quality() == daphne::MEASUREMENT_ERROR) history.reset();
    else {
      const auto bound = history.observe(raw, start, end);
      if (bound && result.last_sample().quality() == daphne::MEASUREMENT_GOOD) {
        need(*bound <= start);
        result.mutable_last_sample()->set_age_bound_quality(daphne::MEASUREMENT_GOOD);
        result.mutable_last_sample()->set_not_before_monotonic_ns(*bound);
        result.mutable_last_sample()->set_maximum_age_ns(end - *bound);
      }
    }
    result.set_observed_monotonic_ns(end); result.set_quality(daphne::MEASUREMENT_GOOD);
    result.set_detail("Pinned-owner service observation, not active-unit health or synchronization. Selected peer is not necessarily the origin of a retained sample. Unavailable sample age is not zero");
  } catch (const std::exception& error) {
    history.reset();
    const auto* system = dynamic_cast<const std::system_error*>(&error);
    const auto code = system ? system->code() : std::error_code{};
    const bool absent = system && (code == std::errc::no_such_file_or_directory || code == std::errc::no_such_process ||
        code == std::errc::permission_denied || code == std::errc::operation_not_permitted || code == std::errc::connection_refused);
    const auto start = result.acquisition_started_monotonic_ns();
    const auto source = result.source();
    result.Clear(); result.set_source(source); result.set_acquisition_started_monotonic_ns(start);
    result.set_quality(absent ? daphne::MEASUREMENT_UNAVAILABLE : daphne::MEASUREMENT_ERROR);
    result.set_detail(absent ? "Timesync service/bus unavailable; no service activation attempted" : "Timesync query, owner, data or acquisition validation failed; private details suppressed");
    result.mutable_last_sample()->set_quality(daphne::MEASUREMENT_UNAVAILABLE);
    result.mutable_last_sample()->set_age_bound_quality(daphne::MEASUREMENT_UNAVAILABLE);
    result.mutable_last_sample()->set_detail("No usable current service observation");
  }
  return result;
}

void add_timesync_status(daphne::SystemStatusSnapshot& status, bool private_details) {
  static std::mutex mutex;
  static TimesyncHistory history;
  std::lock_guard<std::mutex> lock(mutex);
  auto io = make_linux_timesync_io();
  *status.mutable_host_time()->mutable_timesync() = read_timesync(*io, history, private_details);
}
}  // namespace daphne_sc

File: daphne-server/scripts/timesync.py
"""Check timesync1 reporting without promoting history into a current UTC claim."""
import ipaddress
import re

SOURCE = "local system bus:org.freedesktop.timesync1.Manager:GetAll"
META_FIELDS = ("selected_name_present", "selected_address_present", "poll_interval_us", "poll_minimum_us",
               "poll_maximum_us", "root_distance_maximum_us", "processed_packet_count", "frequency_scaled_ppm")
SAMPLE_FIELDS = ("leap", "version", "mode", "stratum", "precision_exponent", "root_delay_us", "root_dispersion_us",
                 "origin_unix_us", "receive_unix_us", "transmit_unix_us", "destination_unix_us", "ignored_spike",
                 "jitter_us", "offset_ns", "round_trip_delay_ns")


def require(value):
    if not value:
        raise RuntimeError("Invalid or stale timesync observation; private details suppressed")


def hostname(value):
    require(0 < len(value) <= 253)
    try:
        ipaddress.ip_address(value)
        return
    except ValueError:
        pass
    require(all(re.fullmatch(r"[A-Za-z0-9](?:[A-Za-z0-9-]{0,61}[A-Za-z0-9])?", label)
                for label in value.rstrip('.').split('.')) and not value.endswith('..'))


def check_timesync(status, high, *, now_monotonic_ns, private_requested=False, require_available=False):
    require(status.HasField("host_time") and status.host_time.HasField("timesync") and now_monotonic_ns > 0)
    item = status.host_time.timesync
    require(item.source == SOURCE and 0 < len(item.detail) <= 512)
    require(item.quality in (high.MEASUREMENT_GOOD, high.MEASUREMENT_UNAVAILABLE, high.MEASUREMENT_ERROR))
    good = item.quality == high.MEASUREMENT_GOOD
    require(good or not require_available)
    for field in META_FIELDS:
        require(item.HasField(field) == good)
    require(item.HasField("last_sample"))
    if good:
        require(item.owner_bracket_verified and bool(re.fullmatch(r"[0-9a-f]{32}", item.bus_id)))
        require(len(item.unique_owner) <= 80 and bool(re.fullmatch(r":[0-9]+(?:\.[0-9]+)+", item.unique_owner)))
        require(0 < item.acquisition_started_monotonic_ns <= item.observed_monotonic_ns <= now_monotonic_ns)
        require(item.observed_monotonic_ns - item.acquisition_started_monotonic_ns <= 2_000_000_000)
        require(now_monotonic_ns - item.observed_monotonic_ns <= 5_000_000_000)
        require(item.details_included == private_requested)
        require(item.HasField("selected_server_name") == (private_requested and item.selected_name_present))
        require(item.HasField("selected_server_address") == (private_requested and item.selected_address_present))
        if item.HasField("selected_server_name"):
            hostname(item.selected_server_name)
        if item.HasField("selected_server_address"):
            require('%' not in item.selected_server_address and len(item.selected_server_address) <= 45)
            try:
                ipaddress.ip_address(item.selected_server_address)
            except ValueError:
                require(False)
        require(0 < item.poll_minimum_us <= item.poll_maximum_us and item.root_distance_maximum_us > 0)
        require(item.poll_interval_us == 0 or item.poll_minimum_us <= item.poll_interval_us <= item.poll_maximum_us)
        if status.host_time.HasField("kernel") and status.host_time.kernel.quality == high.MEASUREMENT_GOOD:
            require(status.host_time.kernel.observed_monotonic_ns <= item.acquisition_started_monotonic_ns)
    else:
        require(not item.observed_monotonic_ns and not item.owner_bracket_verified and not item.details_included)
        require(not item.bus_id and not item.unique_owner)
        require(not item.HasField("selected_server_name") and not item.HasField("selected_server_address"))
        require(item.acquisition_started_monotonic_ns <= now_monotonic_ns)
    sample = item.last_sample
    require(sample.quality in (high.MEASUREMENT_GOOD, high.MEASUREMENT_UNAVAILABLE, high.MEASUREMENT_ERROR))
    require(0 < len(sample.detail) <= 512)
    usable = sample.quality == high.MEASUREMENT_GOOD
    require(not usable or (good and item.processed_packet_count > 0))
    require(sample.quality != high.MEASUREMENT_ERROR or (good and item.processed_packet_count > 0))
    if good and item.processed_packet_count > 0:
        require(sample.quality != high.MEASUREMENT_UNAVAILABLE)
    for field in SAMPLE_FIELDS:
        require(sample.HasField(field) == usable)
    if usable:
        require(sample.leap in range(3) and sample.version in (3, 4) and sample.mode == 4 and 1 <= sample.stratum <= 15)
        require(-128 <= sample.precision_exponent <= 127)
        t1, t2, t3, t4 = (getattr(sample, key + "_unix_us") for key in ("origin", "receive", "transmit", "destination"))
        require(all(0 < t <= ((1 << 63) - 1) // 1000 for t in (t1, t2, t3, t4)))
        delay, offset = (t4 - t1) - (t3 - t2), ((t2 - t1) + (t3 - t4)) * 500
        require(t4 >= t1 and t3 >= t2 and 0 <= delay * 1000 <= (1 << 64) - 1)
        require(-(1 << 63) <= offset <= (1 << 63) - 1)
        require(sample.offset_ns == offset and sample.round_trip_delay_ns == delay * 1000)
    require(sample.age_bound_quality in (high.MEASUREMENT_GOOD, high.MEASUREMENT_UNAVAILABLE))
    bound = sample.age_bound_quality == high.MEASUREMENT_GOOD
    require(not bound or usable)
    require(sample.HasField("not_before_monotonic_ns") == bound and sample.HasField("maximum_age_ns") == bound)
    if bound:
        require(0 < sample.not_before_monotonic_ns <= item.acquisition_started_monotonic_ns)
        require(sample.maximum_age_ns == item.observed_monotonic_ns - sample.not_before_monotonic_ns)
    return {
        "quality": high.MeasurementQuality.Name(item.quality), "service_observed": good,
        "private_details_included": item.details_included,
        "observed_monotonic_ns": item.observed_monotonic_ns,
        **{field: getattr(item, field) if good else None for field in META_FIELDS},
        "sample_quality": high.MeasurementQuality.Name(sample.quality),
        "historical_sample": {field: getattr(sample, field) if usable else None for field in SAMPLE_FIELDS},
        "sample_age_upper_bound_ns": sample.maximum_age_ns if bound else None,
        "scope": "Service and historical NTP data; no selected-peer/sample-origin binding, present offset or verified UTC claim",
    }
